loader/loader.go

```diff
@@ -1,5 +1,165 @@
+// Client to a Database.
 type Loader interface {
+	Run(BlockRange, database.DBTx)
+	Close()
+}
+
+type ClientDBLoaderConstructor func(client client.Client, db database.Database, opts LoaderOptions) (Loader, error)
+
+type ClientDBLoader struct {
+	client     client.Client
+	db         database.Database
+	maxWorkers int
+	msgs       chan LoaderMsg
+}
+
+type LoaderMsg struct {
+	t          LoaderMsgType
+	blockRange BlockRange
+	dbTx       *database.DBTx
+	data       interface{}
+}
+
+// Message type enum
+type LoaderMsgType int
+
+const (
+	Range LoaderMsgType = iota
+	Hashes
+	Blocks
+	BlocksForHeaders
+	BlocksForTxs
+	Headers
+	Txs
+)
+
+type BlockRange struct {
+	startBlockHeight int64
+	endBlockHeight   int64
+}
+
+func NewClientDBLoader(client client.Client, db database.Database, opts LoaderOptions, maxWorkers int) (*ClientDBLoader, error) {
+	// initialize struct
+	msgs := make(chan LoaderMsg)
+	loader := &ClientDBLoader{
+		client:     client,
+		db:         db,
+		msgs:       msgs,
+		maxWorkers: maxWorkers,
+	}
+
+	// start workers
+	for i := 0; i < maxWorkers; i++ {
+		go loader.loaderWorker()
+	}
+
+	return loader, nil
+}
+
+func (loader *ClientDBLoader) Run(blockRange BlockRange, dbTx database.DBTx) {
+	msg := LoaderMsg{
+		Range,
+		blockRange,
+		&dbTx,
+		blockRange,
+	}
+	loader.msgs <- msg
+}
+
+func (loader *ClientDBLoader) loaderWorker() error {
+	for msg := range loader.msgs {
+		switch msg.t {
+		case Range:
+			blockRange := msg.data.(BlockRange)
+			hashes, err := loader.client.GetBlockHashesByRange(blockRange.startBlockHeight, blockRange.endBlockHeight)
+			if err != nil {
+				return err
+			}
+			newMsg := LoaderMsg{Hashes, msg.blockRange, msg.dbTx, hashes}
+			loader.msgs <- newMsg
+
+		case Hashes:
+			hashes := msg.data.([]*chainhash.Hash)
+			blocks, err := loader.client.GetBlocks(hashes)
+			if err != nil {
+				return err
+			}
+			newHeadersMsg := LoaderMsg{BlocksForHeaders, msg.blockRange, msg.dbTx, blocks}
+			loader.msgs <- newHeadersMsg
+			newTxsMsg := LoaderMsg{BlocksForTxs, msg.blockRange, msg.dbTx, blocks}
+			loader.msgs <- newTxsMsg
+
+		case BlocksForHeaders:
+			blocks := msg.data.([]*types.Block)
+			headers, err := BlocksToHeaders(blocks)
+			if err != nil {
+				return err
+			}
+			newMsg := LoaderMsg{Headers, msg.blockRange, msg.dbTx, headers}
+			loader.msgs <- newMsg
+
+		case BlocksForTxs:
+			blocks := msg.data.([]*types.Block)
+			txs, err := BlocksToTxs(blocks)
+			if err != nil {
+				return err
+			}
+			newMsg := LoaderMsg{Txs, msg.blockRange, msg.dbTx, txs}
+			loader.msgs <- newMsg
+
+		case Headers:
+			headers := msg.data.([]*types.BlockHeader)
+			dbTx := nil // will retrieve from somewhere
+			for _, header := range headers {
+				dbTx.AddBlockHeader(header)
+			}
+
+		case Txs:
+			txs := msg.data.([]*types.Transaction)
+			dbTx := nil
+			for _, tx := range txs {
+				dbTx.AddTransaction(tx)
+			}
+
+		default:
+			return fmt.Errorf("Unknown LoaderMsgType %d", msg.t)
+		}
```

Comment:
This all looks pretty good and I don't see why it wouldn't work, but maybe consider a different approach where the pipeline is just a series of channels so that we do not have to handle different types of LoaderMsgs explicitly. So like there would be the `blockRangeCh` which then passes hashes to the `blockHashCh` to `blockHeadersCh` and `txsCh`. Each channel would be handled in its own goroutine.

Also, consider how maybe each process/channel should only be handled by one goroutine.  Blockchains are inherently sequential so when transferring data to another database, it is unlikely that we will be able to store it out of order. We probably can when loading all the historical data into CSVs, but definitely not when doing INSERTS/UPSERTS on a live database. So if there are more workers than MsgTypes in the pipeline, one batch of block ranges may get ahead of a past block range and be committed to the database first, referencing data that may or may not exist yet.

Reply:
> Also, consider how maybe each process/channel should only be handled by one goroutine. Blockchains are inherently sequential so when transferring data to another database, it is unlikely that we will be able to store it out of order. We probably can when loading all the historical data into CSVs, but definitely not when doing INSERTS/UPSERTS on a live database. So if there are more workers than MsgTypes in the pipeline, one batch of block ranges may get ahead of a past block range and be committed to the database first, referencing data that may or may not exist yet.

Originally had the many-channel design in mind but opted for the current one because I thought it'd be simpler with a single channel and I didn't have a reason for disallowing out of order execution. This makes a lot of sense however. Will change.